src/Glpi/Controller/GenericFormController.php

```diff
@@ -0,0 +1,195 @@
+ * ---------------------------------------------------------------------
+ */
+
+namespace Glpi\Controller;
+
+use CommonDBTM;
+use Html;
+use Glpi\Event;
+use Glpi\Exception\Http\AccessDeniedHttpException;
+use Glpi\Exception\Http\BadRequestHttpException;
+use Glpi\Exception\Http\NotFoundHttpException;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Routing\Attribute\Route;
+
+class GenericFormController extends AbstractController
+{
+    private const SUPPORTED_ACTIONS = [
+        'add',
+        'delete',
+        'restore',
+        'purge',
+        'update',
+        'unglobalize',
+    ];
+
+    #[Route("/{class}/Form", name: "glpi_itemtype_form", priority: -1)]
+    public function __invoke(Request $request): Response
+    {
+        $class = $request->attributes->getString('class');
+
+        $this->checkIsValidClass($class);
+
+        /** @var class-string<CommonDBTM> $class */
+
+        if (!$class::canView()) {
+            throw new AccessDeniedHttpException();
+        }
+
+        $form_action = $this->getFormAction($request, $class);
+
+        if (!$form_action) {
+            throw new BadRequestHttpException();
+        }
+
+        return $this->handleFormAction($request, $form_action, $class);
+    }
+
+    private function checkIsValidClass(string $class): void
+    {
+        if (!$class) {
+            throw new BadRequestHttpException('The "class" attribute is mandatory for dropdown routes.');
+        }
+
+        if (!\class_exists($class)) {
+            throw new BadRequestHttpException(\sprintf("Class \"%s\" does not exist.", $class));
+        }
+
+        if (!\is_subclass_of($class, CommonDBTM::class)) {
+            throw new BadRequestHttpException(\sprintf("Class \"%s\" is not a DB object.", $class));
+        }
+    }
+
+    /**
+     * @param class-string<CommonDBTM> $class
+     */
+    private function handleFormAction(Request $request, string $form_action, string $class): Response
+    {
+        $id = $request->query->get('id', -1);
+        $post_data = $request->request->all();
+
+        /* @var CommonDBTM $object */
+        $object = new $class();
+
+        if (!$object::isNewID($id) && !$object->getFromDB($id)) {
+            throw new NotFoundHttpException();
+        }
+
+        // Special case for GET
+        if ($form_action === 'get') {
+            if (!$object->can($id, READ)) {
+                throw new AccessDeniedHttpException();
+            }
+            return $this->displayForm($object, $request);
+        }
+
+        // Permissions
+        $can_do_action = match ($form_action) {
+            'add' => $object->can($id, CREATE, $post_data),
+            'delete', 'restore' => $object->can($id, DELETE, $post_data),
+            'purge' => $object->can($id, PURGE, $post_data),
+            default => $object->can($id, UPDATE, $post_data),
+        };
```

Comment:
This will check for `update` right when trying to restore an item.
The old behavior was to check `delete`:

![image](https://github.com/user-attachments/assets/e0fac721-2c33-46e0-a45f-faa1def68106)

To avoid this kind of mistake I think it would be better to do something similar to the `match` block you did just below:
* Have a specific case for each possible action
* Exception on default

Reply:
Nevermind, missed the `, restore`.